File: AspNetCore.ApiGateway.Client/ApiGatewayClient.cs
using Microsoft.AspNetCore.JsonPatch;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace AspNetCore.ApiGateway.Client
{
    public class ApiGatewayClient : IApiGatewayClient
    {
        private readonly HttpClient _httpClient;
        private readonly ApiGatewayClientSettings _settings;

        public ApiGatewayClient(IHttpService httpService, ApiGatewayClientSettings settings)
        {
            _httpClient = httpService.HttpClient;
            _settings = settings;
        }

        public async Task<TResponse> GetAsync<TResponse>(ApiGatewayParameters parameters)
        {
            var gatewayUrl = UrlCombine(_settings.ApiGatewayBaseUrl, "api/Gateway", parameters.ApiKey, parameters.RouteKey);
            gatewayUrl = $"{gatewayUrl}?parameters={WebUtility.UrlEncode(parameters.Parameters ?? String.Empty)}";

            _httpClient.AddHeaders(parameters);

            var response = await _httpClient.GetAsync(gatewayUrl);

            response.EnsureSuccessStatusCode();

            var returnedContent = await response.Content.ReadAsStringAsync();

            return !string.IsNullOrEmpty(returnedContent) ? JsonSerializer.Deserialize<TResponse>(returnedContent) : default(TResponse);
        }

        public async Task<TResponse> PostAsync<TPayload, TResponse>(ApiGatewayParameters parameters, TPayload data)
        {
            var gatewayUrl = UrlCombine(_settings.ApiGatewayBaseUrl, "api/Gateway", parameters.ApiKey, parameters.RouteKey);
            gatewayUrl = $"{gatewayUrl}?parameters={WebUtility.UrlEncode(parameters.Parameters ?? String.Empty)}";

            _httpClient.AddHeaders(parameters);

            var content = new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json");

            var response = await _httpClient.PostAsync(gatewayUrl, content);

            response.EnsureSuccessStatusCode();

            var returnedContent = await response.Content.ReadAsStringAsync();

            return !string.IsNullOrEmpty(returnedContent) ? JsonSerializer.Deserialize<TResponse>(returnedContent) : default(TResponse);
        }

        public async Task PutAsync<TPayload>(ApiGatewayParameters parameters, TPayload data)
        {
            await this.PutAsync<TPayload, string>(parameters, data);
        }

        public async Task<TResponse> PutAsync<TPayload, TResponse>(ApiGatewayParameters parameters, TPayload data)
        {
            var gatewayUrl = UrlCombine(_settings.ApiGatewayBaseUrl, "api/Gateway", parameters.ApiKey, parameters.RouteKey);
            gatewayUrl = $"{gatewayUrl}?parameters={WebUtility.UrlEncode(parameters.Parameters ?? String.Empty)}";

            _httpClient.AddHeaders(parameters);

            var content = new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json");
            content.Headers.ContentType = new MediaTypeHeaderValue("application/json");

            var response = await _httpClient.PutAsync(gatewayUrl, content);

            response.EnsureSuccessStatusCode();

            var returnedContent = await response.Content.ReadAsStringAsync();

            return !string.IsNullOrEmpty(returnedContent) ? JsonSerializer.Deserialize<TResponse>(returnedContent) : default(TResponse);
        }

        public async Task PatchAsync<TPayload>(ApiGatewayParameters parameters, JsonPatchDocument<TPayload> data)
            where TPayload : class
        {
            await this.PatchAsync<TPayload, string>(parameters, data);
        }

        public async Task<TResponse> PatchAsync<TPayload, TResponse>(ApiGatewayParameters parameters, JsonPatchDocument<TPayload> data)
            where TPayload : class
        {
            var gatewayUrl = UrlCombine(_settings.ApiGatewayBaseUrl, "api/Gateway", parameters.ApiKey, parameters.RouteKey);
            gatewayUrl = $"{gatewayUrl}?parameters={WebUtility.UrlEncode(parameters.Parameters ?? String.Empty)}";

            _httpClient.AddHeaders(parameters);

            var content = new StringContent(JsonSerializer.Serialize(data.Operations), Encoding.UTF8, "application/json-patch+json");
            content.Headers.ContentType = new MediaTypeHeaderValue("application/json-patch+json");

            var method = "PATCH";
            var httpVerb = new HttpMethod(method);

            var httprequest = new HttpRequestMessage
            {
                RequestUri = new Uri(gatewayUrl),
                Content = content,
                Method = httpVerb
            };

            var response = await _httpClient.SendAsync(httprequest);

            response.EnsureSuccessStatusCode();

            var returnedContent = await response.Content.ReadAsStringAsync();

            return !string.IsNullOrEmpty(returnedContent) ? JsonSerializer.Deserialize<TResponse>(returnedContent) : default(TResponse);
        }

        public async Task DeleteAsync(ApiGatewayParameters parameters)
        {
            await this.DeleteAsync<string>(parameters);
        }

        public async Task<TResponse> DeleteAsync<TResponse>(ApiGatewayParameters parameters)
        {
            var gatewayUrl = UrlCombine(_settings.ApiGatewayBaseUrl, "api/Gateway", parameters.ApiKey, parameters.RouteKey);
            gatewayUrl = $"{gatewayUrl}?parameters={WebUtility.UrlEncode(parameters.Parameters ?? String.Empty)}";

            _httpClient.AddHeaders(parameters);

            var response = await _httpClient.DeleteAsync(gatewayUrl);

            response.EnsureSuccessStatusCode();

            var returnedContent = await response.Content.ReadAsStringAsync();

            return !string.IsNullOrEmpty(returnedContent) ? JsonSerializer.Deserialize<TResponse>(returnedContent) : default(TResponse);
        }

        public async Task<IEnumerable<Orchestration>> GetOrchestrationAsync(ApiGatewayParameters parameters)
        {
            var gatewayUrl = UrlCombine(_settings.ApiGatewayBaseUrl, $"api/Gateway/orchestration?api={parameters.ApiKey}&key={parameters.RouteKey}");

            _httpClient.AddHeaders(parameters);

            var response = await _httpClient.GetAsync(gatewayUrl);

            response.EnsureSuccessStatusCode();

            var content = await response.Content.ReadAsStringAsync();

            var orchs = JArray.Parse(content);

            var orchestrations = new List<Orchestration>();

            orchs.AsEnumerable().ToList().ForEach(item =>
            {
                var type = item.SelectToken("orchestrationType");
                var orchestrationType = (OrchestationType)Enum.Parse(typeof(OrchestationType), type.ToString());

                switch (orchestrationType)
                {
                    case OrchestationType.Api:
                        orchestrations.Add(item.ToObject<ApiOrchestration>());
                        break;
                    case OrchestationType.Hub:
                        orchestrations.Add(item.ToObject<HubOrchestration>());
                        break;
                    case OrchestationType.EventSource:
                        orchestrations.Add(item.ToObject<EventSourceOrchestration>());
                        break;
                }
            });

            return orchestrations;
        }

        private static string UrlCombine(string baseUrl, params string[] segments)
                    => string.Join("/", new[] { baseUrl.TrimEnd('/') }.Concat(segments.Select(s => s.Trim('/'))));
    }
}

File: ts-aspnetcore-apigateway-client/ApiGatewayClient.ts
import fetch from 'node-fetch';
import path from 'path';
import fs from 'fs';
import { ApiGatewayClientSettings } from './ApiGatewayClientSettings';
import { ApiGatewayParameters } from "./ApiGatewayParameters";
import { IApiGatewayClient } from "./IApiGatewayClient";
import { JsonPatchOperation } from './JsonPatch';
import { Orchestration } from './Orchestration';

export class ApiGatewayClient implements IApiGatewayClient {

    _settings: ApiGatewayClientSettings;
    _httpsAgent: any;

    constructor(settings: ApiGatewayClientSettings) {
        this._settings = settings;

        if (this._settings.UseHttps) {
            const https = require('https');

            if (this._settings.IsDEVMode) {
                this._httpsAgent = new https.Agent({
                    rejectUnauthorized: false,
                    requestCert: true,
                    agent: false            
                });
            }
            else {
                if (this._settings.UseCertificate) {
                    const options = {
                        cert: fs.readFileSync(
                          path.resolve(__dirname, this._settings.HttpsSettings?.PfxPath!),
                          `utf-8`,
                        ),
                        key: fs.readFileSync(
                          path.resolve(__dirname, this._settings.HttpsSettings?.PrivateKeyPath!),
                          'utf-8',
                        ),
                        passphrase:
                            this._settings.HttpsSettings?.Passphrase!,
                    
                        rejectUnauthorized: true,
                    
                        keepAlive: true,
                      };
                      this._httpsAgent = new https.Agent(options);
                }                
            }            
        }      
    }

    async GetAsync<TResponse>(parameters: ApiGatewayParameters): Promise<TResponse> {
        let gatewayUrl = `${this._settings.ApiGatewayBaseUrl}/api/Gateway/${parameters.Api}/${parameters.Key}?parameters=${parameters.Parameters??""}`;    
        
        var options = this._settings.IsDEVMode || this._settings.UseCertificate ? {method: 'GET', agent: this._httpsAgent} : {method: 'GET'};
        
        const response = await fetch(gatewayUrl, options);
        const res = await response.json();

        return <TResponse> res;        
    }

    async PostAsync<TPayload, TResponse>(parameters: ApiGatewayParameters, data: TPayload): Promise<TResponse> {
        let gatewayUrl = `${this._settings.ApiGatewayBaseUrl}/api/Gateway/${parameters.Api}/${parameters.Key}?parameters=${parameters.Parameters??""}`;
        
        let headers = { 'Content-Type': 'application/json' };
        let body = JSON.stringify(data);

        var options = this._settings.IsDEVMode || this._settings.UseCertificate ? {method: 'POST', body: body, headers: headers, agent: this._httpsAgent} 
                                                                                : {method: 'POST', body: body, headers: headers};
        
        const response = await fetch(gatewayUrl, options);
        const res = await response.json();

        return <TResponse> res;        
    }
    
    async PutAsync<TPayload, TResponse>(parameters: ApiGatewayParameters, data: TPayload): Promise<TResponse> {
        let gatewayUrl = `${this._settings.ApiGatewayBaseUrl}/api/Gateway/${parameters.Api}/${parameters.Key}?parameters=${parameters.Parameters??""}`;
        
        let headers = { 'Content-Type': 'application/json' };
        let body = JSON.stringify(data);

        var options = this._settings.IsDEVMode || this._settings.UseCertificate ? {method: 'PUT', body: body, headers: headers, agent: this._httpsAgent} 
                                                                                : {method: 'PUT', body: body, headers: headers};
        
        const response = await fetch(gatewayUrl, options);

        if (response.ok && response.bodyUsed) {
            const res = await response.json();

            return <TResponse> res;
        }
                
        return <TResponse>{};
    }    
    
    async PatchAsync<TResponse>(parameters: ApiGatewayParameters, data: JsonPatchOperation[]): Promise<TResponse> {
        let gatewayUrl = `${this._settings.ApiGatewayBaseUrl}/api/Gateway/${parameters.Api}/${parameters.Key}?parameters=${parameters.Parameters??""}`;
        
        let headers = { 'Content-Type': 'application/json-patch+json' };
        let body = JSON.stringify(data);

        var options = this._settings.IsDEVMode || this._settings.UseCertificate ? {method: 'PATCH', body: body, headers: headers, agent: this._httpsAgent} 
                                                                                : {method: 'PATCH', body: body, headers: headers};
        
        const response = await fetch(gatewayUrl, options);

        if (response.ok) {
            const res = await response.json();

            return <TResponse> res;
        }
                
        return <TResponse>{};
    } 

    async DeleteAsync<TResponse>(parameters: ApiGatewayParameters): Promise<TResponse> {
        let gatewayUrl = `${this._settings.ApiGatewayBaseUrl}/api/Gateway/${parameters.Api}/${parameters.Key}?parameters=${parameters.Parameters??""}`;     
        
        var options = this._settings.IsDEVMode || this._settings.UseCertificate ? {method: 'DELETE', agent: this._httpsAgent} : {method: 'DELETE'};
        
        const response = await fetch(gatewayUrl, options);
        
        if (response.ok && response.bodyUsed) {
            const res = await response.json();

            return <TResponse> res;
        }

        return <TResponse>{};        
    }

    async GetOrchestrationAsync(parameters: ApiGatewayParameters): Promise<Orchestration[]> {
        let gatewayUrl = `${this._settings.ApiGatewayBaseUrl}/api/Gateway/orchestration?api=${parameters.Api}&key=${parameters.Key}`;   
        
        var options = this._settings.IsDEVMode || this._settings.UseCertificate ? {method: 'GET', agent: this._httpsAgent} : {method: 'GET'};
        
        const response = await fetch(gatewayUrl, options);
        const res = await response.json();

        return <Orchestration[]> res;        
    }    
}